mqtt/mqtt-bridge/src/persist/storage/ring_buffer/block.rs

```diff
@@ -0,0 +1,201 @@
+use getset::{Getters, MutGetters, Setters};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::hash_map::DefaultHasher,
+    fmt::Debug,
+    hash::{Hash, Hasher},
+};
+
+use crate::persist::storage::serialize::binary_serialize;
+
+use super::{error::BlockError, RingBufferResult};
+
+/// A utility fn to calculate any entity that derives Hash.
+pub(crate) fn calculate_hash<T>(t: &T) -> u64
+where
+    T: Hash + ?Sized,
+{
+    let mut hasher = DefaultHasher::new();
+    t.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// + ------------+------+---------+
+/// | BlockHeader | hash | data... |
+/// +-------------+------+---------+
+///
+/// The BlockHeader contains attributes meaningful for data
+/// storage and validation.
+#[derive(Copy, Clone, Debug, Deserialize, Getters, Setters, Hash, Serialize)]
+#[repr(C)]
+pub(crate) struct BlockHeader {
+    #[getset(get = "pub")]
+    data_size: usize,
+    #[getset(get = "pub")]
+    index: usize,
+    #[getset(get = "pub")]
+    data_hash: u64,
+    #[getset(get = "pub", set = "pub")]
+    should_not_overwrite: bool,
+}
+
+impl BlockHeader {
+    pub fn new(data_size: usize, index: usize, data_hash: u64) -> Self {
+        Self {
+            data_size,
+            index,
+            data_hash,
+            should_not_overwrite: true,
+        }
+    }
+}
+
+/// + --------------------+---------+
+/// | BlockHeaderWithHash | data... |
+/// +---------------------+---------+
+///
+/// The BlockHeaderWithHash is a wrapper over the BlockHeader
+/// but also contains a hash over the header for validation.
+#[derive(Copy, Clone, Debug, Deserialize, Getters, MutGetters, Hash, Serialize)]
+#[repr(C)]
+pub(crate) struct BlockHeaderWithHash {
+    #[getset(get = "pub", get_mut = "pub")]
+    inner: BlockHeader,
+    #[getset(get = "pub")]
+    block_hash: u64,
+}
+
+impl BlockHeaderWithHash {
+    pub fn new(data_size: usize, index: usize, data_hash: u64) -> Self {
+        let header = BlockHeader::new(data_size, index, data_hash);
+        let header_hash = calculate_hash(&header);
+        Self {
+            inner: header,
+            block_hash: header_hash,
+        }
+    }
+}
+
+/// A typed representation of byte data that should always
+/// follow a BlockHeaderWithHash.
+#[derive(Clone, Debug, Deserialize, Getters, Hash, Serialize)]
+#[repr(C)]
+pub(crate) struct Data {
+    #[getset(get = "pub")]
+    inner: Vec<u8>,
+}
+
+impl Data {
+    pub fn new(data: Vec<u8>) -> Self {
+        Self { inner: data }
+    }
+}
+
+pub(crate) fn validate(block: &BlockHeaderWithHash, data: &Data) -> RingBufferResult<()> {
```

Comment:
Let's make sure that the validation of block looks 100% correct :) So if I can get some eyeballs looking over this that will be very helpful :)